content/blog/2025-12-16-tiling-fusion-search.md

```diff
@@ -0,0 +1,135 @@
+        for n0 in [0, N0):
+            ...
+                C = einsum0(A, B)
+                E = einsum1(C, D)
+                H = einsum2(E, F, G)
+                ...
+
+However, this is not optimal. Sometimes the einsums should be fused on different levels; sometimes tensors other than the passed tensor should take the outermost blocking level. Unfortunately, there does not seem to be any simple way to modify the above program to fix this. We had to go with a totally new approach.
+
+This "full-tree" approach is [here](https://github.com/natetyoung/tensor-program-dse/blob/main/full_tree_cpsat97.py). Rather than building orders of tiling levels in multiple einsums individually, each creating their own buckets, this CP-SAT formulation creates a tree structure among all the tensors in the entire program. Instead of constraints on "factors outside this tiling level," this program has constraints on "factors which are ancestors of this tiling level in the tree," and so on. In effect, this CP-SAT program synthesizes a restricted form of a [schedule tree](https://acohen.gitlabpages.inria.fr/impact/impact2014/papers/impact2014-verdoolaege.pdf) to create an optimal tiling and fusion for an input chain of einsums.
+
+For a 2-GEMM example given in the Orojenesis paper (`32k x 4k x 16k` and `32k x 16k x 4k`), it takes 0.76 seconds on my laptop.
+
+# The hardest part
+
+Constraint programs in CP-SAT are fairly difficult to debug, since if the program turns out to be infeasible there's very little way of figuring out why, and if an expected output isn't found, it can be difficult to figure out exactly which constraints it violated. For this reason, every CP-SAT formulation took some time to get right, but by far the hardest was the final "full-tree" program, especially the part of the program to estimate which buffers need to coexist with which other buffers, so we could get a total necessary cache size. Also, we found a bug in recent versions of CP-SAT and had to revert to an earlier version.
+
+# Success
+
+We evaluated our project on some real-world tensor operations against -O3 without vectorization or loop unrolling. We follow the evaluation criteria used in Orojenesis by generating some C code with dummy values, which models a chain of 3 einsums (Q_proj, QK, QKV) from an inference pass through a GPT-style block. Notice that the output of each einsum is fed as input to the next, allowing for loop fusion. To vary the sizes of these operations, we adjusted the corresponding "sequence length" and "model dimensions" used to generate the einsums in the test code, and used a spread of different sizes: (`seq`, `dim`) = (`32, 128`), (`64`, `256`), (`128`, `512`), (`1024`, `4096`). We refer to them as `tiny`, `small`, `med`, and `large`, respectively. Note that the dimensions under `large` actually correspond to realistic dimensions used in the [Cerebras GPT-3-6.7B](https://huggingface.co/cerebras/Cerebras-GPT-6.7B) block.
```

Comment:
I think you haven't mentioned yet that your compiler can generate C code that implements a given schedule for a given einsum. (Assuming that it can? And that you're not hand-writing each program according to the solver output.) Can you describe that a bit earlier, in your implementation section?